File: backend/src/main/kotlin/com/example/backend/services/UserService.kt
package com.example.backend.services

import com.example.backend.dtos.UserNewDto
import com.example.backend.models.Employe
import com.example.backend.models.User
import com.example.backend.repositories.UserRepository
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.*

@Service
class UserService(
  private val userRepository: UserRepository
) {

  fun getAllUsers(): List<User> {
    return userRepository.findAll()
  }

  fun getUserById(id: Long): Optional<User> {
    return userRepository.findById(id.toInt())
  }

  fun createUser(user: User): User {
//    user.createdDate = LocalDateTime.now()
    return userRepository.save(user)
  }

  fun updateUser(id: Long, updatedUser: UserNewDto): User {

    if (id== 0.0.toLong()){
      var user = User().apply {
        this.nom = updatedUser.nom
        this.prenom = updatedUser.prenom
        this.email = updatedUser.email
        this.password = null
        this.username = null
        this.fonction = null
        this.role = null
        this.telephone = updatedUser.telephone
        this.reduction = updatedUser.reduction!!.toInt()
        this.reductionMax = updatedUser.reductionMax!!.toInt()
        this. supprimer = 0
      }
      return userRepository.save(user)
    }
    else {
      val existingUser = userRepository.findById(id.toInt())
        .orElseThrow { IllegalArgumentException("User with ID $id not found") }
      existingUser.nom = updatedUser.nom
      existingUser.prenom = updatedUser.prenom
      existingUser.email = updatedUser.email
      existingUser.password = null
      existingUser.username = null
      existingUser.fonction = null
      existingUser.role = null
      existingUser.telephone = updatedUser.telephone
      existingUser.reduction = updatedUser.reduction!!.toInt()
      existingUser.reductionMax = updatedUser.reductionMax!!.toInt()
      existingUser. supprimer = 0
      return userRepository.save(existingUser)
    }

  }

  fun deleteUser(id: Long) {
    if (!userRepository.existsById(id.toInt())) {
      throw IllegalArgumentException("User with ID $id not found")
    }
//    userRepository.deleteById(id.toInt())
    var user = userRepository.findById(id.toInt()).get()
    user.supprimer=1
    userRepository.save(user)
  }


  fun getCurrentUser(): User? {
    val auth = SecurityContextHolder.getContext().authentication
    val username = when (val principal = auth?.principal) {
      is UserDetails -> principal.username
      is String -> principal
      else -> null
    }
    return username?.let { userRepository.findByEmail(it) }
  }
}

File: backend_py/pharmaPython/app/services/user_service.py
# services/user_service.py
from sqlalchemy.orm import Session

class UserService:
  def __init__(self, db: Session, user_repo):
    self.db = db
    self.user_repo = user_repo

  def get_all_users(self):
    return self.user_repo.find_all()

  def get_user_by_id(self, id_: int):
    return self.user_repo.find_by_id(id_)

  def create_user(self, user):
    return self.user_repo.save(user)

  def update_user(self, id_: int, dto):
    if id_ == 0:
      user = {
        "nom": dto.nom, "prenom": dto.prenom, "email": dto.email,
        "username": None, "password": None, "role": None,
        "telephone": dto.telephone,
        "reduction": dto.reduction, "reductionMax": dto.reductionMax,
        "supprimer": 0
      }
      return self.user_repo.save(user)
    else:
      u = self.user_repo.find_by_id(id_)
      if not u:
        raise ValueError(f"User {id_} not found")
      u.nom, u.prenom, u.email = dto.nom, dto.prenom, dto.email
      u.username, u.password, u.role = None, None, None
      u.telephone = dto.telephone
      u.reduction = dto.reduction
      u.reductionMax = dto.reductionMax
      u.supprimer = 0
      return self.user_repo.save(u)

  def delete_user(self, id_: int):
    u = self.user_repo.find_by_id(id_)
    if not u:
      raise ValueError(f"User {id_} not found")
    u.supprimer = 1
    return self.user_repo.save(u)

  def get_current_user(self, username: str):
    return self.user_repo.find_by_email(username)
